GS:Mac: Add ffmpeg

TellowKrinkle opened the pull request:
### Description of Changes
- Adds a minimal (~3MB) copy of ffmpeg to Mac builds (includes a few small lossless codecs and the interface to macOS's system encoding libraries)
- Fixes a bug where s16 codecs would get their audio mangled
- Fixes a bug where setting a pixel format with `pixel_format=fmt` in the extra arguments would break things

### Rationale behind Changes
Capture support on macOS

### Suggested Testing Steps
Try recording things on macOS

stenzek commented on pcsx2/GS/GSCapture.cpp:
```diff
@@ -545,7 +545,7 @@ bool GSCapture::BeginCapture(float fps, GSVector2i recommendedResolution, float
 		}
 
-		s_converted_video_frame->format = sw_pix_fmt;
+		s_converted_video_frame->format = s_video_codec_context->pix_fmt;
```
Are you sure about this one? For hardware encoders, the context pixel format may be different, and we want to convert from RGB->YUV (or whatever sw_pix_fmt is), then to hw.

TellowKrinkle replied:
I'm not sure about this one, none of the macOS encoders return something that uses an AVCodecHWConfig

Things I can confirm:
- The macOS hardware encoder, h264_videotoolbox, does not have a AVCodecHWConfig, but offers the pixel formats `AV_PIX_FMT_VIDEOTOOLBOX` (hw pixel format), `AV_PIX_FMT_NV12`, and `AV_PIX_FMT_YUV420P`.  If we request `AV_PIX_FMT_YUV420P`, the resulting format on `s_video_codec_context` is `AV_PIX_FMT_YUV420P`.
- If you select FFV1 and add `pixel_format=yuv444p` to the extra arguments, you get broken output if you try to feed it yuv420p frames.

TellowKrinkle replied:
Ahh I see where we write the hw format in
I changed it to only use that one if the user specified a `pixel_format` argument

JordanTheToaster commented:
Tested on Windows and it works correctly it also fixes recording with FLAC cannot test on Mac OS.

DigitalMajestic commented:
I'm getting "failed to find video encoder" errors on Mac regardless of what output file format I try (MOV, MKV, etc). Also, weirdly, it wanted to access my calendar. That's probably just Mac being stupid, I don't think we touch that on purpose.

<img width="1440" alt="Screenshot 2023-06-27 at 12 28 22 PM" src="https://github.com/PCSX2/pcsx2/assets/40978750/5d645914-2346-4e0c-a923-fab6be824255">

<img width="1440" alt="Screenshot 2023-06-27 at 12 28 44 PM" src="https://github.com/PCSX2/pcsx2/assets/40978750/a4de76e5-1b07-46c4-9f3c-08fefa7a630d">

TellowKrinkle commented:
> I'm getting "failed to find video encoder" errors on Mac regardless of what output file format I try (MOV, MKV, etc)

Should be fixed now

> Also, weirdly, it wanted to access my calendar. That's probably just Mac being stupid, I don't think we touch that on purpose.

Qt's file open/save dialog was touching it.  Fixed in #9104 which I've rebased this on top of

DigitalMajestic commented:
Permissions spam is fixed. Default encoders failing is not. Manually selecting the proper encoders works as expected.

TellowKrinkle commented:
> Permissions spam is fixed. Default encoders failing is not. Manually selecting the proper encoders works as expected.

Try now

stenzek commented on pcsx2/GS/GSCapture.cpp:
```diff
@@ -536,6 +545,10 @@ bool GSCapture::BeginCapture(float fps, GSVector2i recommendedResolution, float
 		}
 
+		// If the user overrode the pixel format, get that now
+		if (has_pixel_format_override)
```
I'm fairly sure this is going to break hardware encoding, since the sw pixel format gets set above. Maybe move this (and the arguments parse) up to before line 458.

TellowKrinkle replied:
The arguments parse just converts the string to a dictionary.  The values don't get put into the context until the `avcodec_open2` call.  I don't think there's much we can do about hw encoders if the user tries to override the pixel format, unless we want to parse arguments ourselves.  I'm not too worried about it though, since I think the main reason you'd want to change pixel formats is if you want a lossless codec for something, and don't want chroma subsampling.

stenzek replied:
Right. But check https://github.com/PCSX2/pcsx2/blob/d0ce4c52b0f95c41083537ed13449233e49a6af7/pcsx2/GS/GSCapture.cpp#L489-L490

If the user overrode the pixel format, then the context would be using one format, with the hardware context using another.
